### ISSUE TITLE
heavy io doing glob('**') in /web/simple for global index

### ISSUE THREAD

**Issue by TechCiel** [pypa/bandersnatch#1337]
https://github.com/pypa/bandersnatch/blob/16137fd3861dd41cd34c060989140edeb00f8e43/src/bandersnatch/simple.py#L78-L89

The function shown above generate a package list for the global indexes i.e. `/web/simple/index{.html,_v1.html,_v1.json}`.

According to the [pathlib.Path.glob document](https://docs.python.org/3/library/pathlib.html#pathlib.Path.glob):

> Note: Using the “**” pattern in large directory trees may consume an inordinate amount of time.

This function is extremely slow (minutes) for a frequent mirror on HDD. `bandersnatch` now do this slow glob in a directory having over 400,000 sub-directories even for just 1 package update. Some tracing shows that, each sub-directory (and possibly the `index.html` file) is `stat()`ed during this process.

It holds in my mirror that, **each sub-directory in `/web/simple` contains a `index.html`**, so I think it's sufficient to just `readdir(simple_dir)` for a comprehensive package list here. So I tried to replace the set comprehension to the one below:

```python
{
    x.name
    for x in simple_path.iterdir()
    if 'index.' not in x.name
}
```

After this change, generating global index now takes only seconds. But I'm still not sure if there's any other considerations or legacy problems preventing doing this, hence this issue instead of a direct PR is opened.

**Comment by cooperlees** [comment]
Howdy,

I am not sure of any short fall moving to a faster readdir. This seems a great enhancement. I just believe we need to test it works with `hash-index = true` sharding set on in config, which I think it will. The simple index dist are just one level deeper in the directory structure to help with the large amount of directories in simple. I will accept the PR is we can prove no regressions here.

It would be nice if we could come up with a unittest way to ensure we don't regress here. i feel we have some coverage, but nothing directly. So if you have time to workout if we do have coverage in tests or add some dedicated ones I'd be appreciative!

**Comment by TechCiel** [comment]
fixed in #1340
